// @uploadcare/react-widget/en/dialog-08094098.js
import { c as _objectSpread2, b as _slicedToArray } from './index-c521bff0.js';
import React, { useReducer, useRef, useEffect } from 'react';
import '@uploadcare/client-suspense';
import { u as useCustomTabs } from './use-custom-tabs-7f831c88.js';
import 'react-fast-compare';
import uploadcare from 'uploadcare-widget/uploadcare.lang.en';

var merge = function merge(state, newState) {
  return _objectSpread2(_objectSpread2({}, state), newState);
};

var useState = function useState(initial) {
  return useReducer(merge, initial || {});
};

var containerStyles = {
  height: '500px',
  display: 'flex',
  flexDirection: 'column',
  alignItems: 'center',
  justifyContent: 'center'
};

var Progress = function Progress(_ref) {
  var hidden = _ref.hidden,
      value = _ref.value;
  var ref = useRef(null);
  var renderer = useRef(null);
  useEffect(function () {
    renderer.current = new uploadcare.Circle(ref.current).renderer;
  }, []);
  useEffect(function () {
    renderer.current.setValue(value);
  }, [value]);
  return /*#__PURE__*/React.createElement("div", {
    hidden: hidden,
    ref: ref
  });
};

var useDialog = function useDialog(props, uploadcare) {
  var _useState = useState({
    opened: true,
    file: null,
    progress: null
  }),
      _useState2 = _slicedToArray(_useState, 2),
      state = _useState2[0],
      setState = _useState2[1];

  var customTabs = props.customTabs,
      tabsCss = props.tabsCss;
  useCustomTabs(customTabs, uploadcare);
  var panelContainer = useRef(null);
  var panelInstance = useRef(null);
  useEffect(function () {
    if (state.opened) {
      if (uploadcare && tabsCss && typeof tabsCss === 'string') {
        if (tabsCss.indexOf('https://') === 0) {
          uploadcare.tabsCss.addUrl(tabsCss);
        } else {
          uploadcare.tabsCss.addStyle(tabsCss);
        }
      }

      panelInstance.current && panelInstance.current.reject();
      panelInstance.current = uploadcare.openPanel(panelContainer.current, state.file ? uploadcare.fileFrom('uploaded', state.file.uuid, props) : null, props);
      setState({
        file: null
      });
      panelInstance.current.done(function (file) {
        setState({
          opened: false
        });
        file.progress(function (state) {
          return setState({
            progress: state.progress
          });
        });
        file.done(function (file) {
          return setState({
            file: file
          });
        });
      });
    } // eslint-disable-next-line react-hooks/exhaustive-deps

  }, [uploadcare, props, setState, state.opened]);
  useEffect(function () {
    return function () {
      return panelInstance.current && panelInstance.current.reject();
    };
  }, []);
  return [state, setState, panelContainer];
};

var Dialog = function Dialog(props) {
  var _useDialog = useDialog(props, uploadcare),
      _useDialog2 = _slicedToArray(_useDialog, 3),
      _useDialog2$ = _useDialog2[0],
      opened = _useDialog2$.opened,
      file = _useDialog2$.file,
      progress = _useDialog2$.progress,
      setState = _useDialog2[1],
      containerRef = _useDialog2[2];

  var readyState = file && !opened;
  var progressState = !file && !opened && progress !== null;
  return /*#__PURE__*/React.createElement("div", {
    id: props.id,
    style: containerStyles
  }, /*#__PURE__*/React.createElement("div", {
    ref: containerRef
  }), /*#__PURE__*/React.createElement("span", {
    hidden: !readyState
  }, /*#__PURE__*/React.createElement("button", {
    className: "uploadcare--button uploadcare--button_primary",
    onClick: function onClick() {
      return setState({
        opened: true,
        sourse: 'natural'
      });
    }
  }, "Open panel")), /*#__PURE__*/React.createElement("span", {
    hidden: !readyState
  }, "File uuid: "), /*#__PURE__*/React.createElement("span", {
    hidden: !readyState
  }, file && file.uuid), /*#__PURE__*/React.createElement(Progress, {
    hidden: !progressState,
    value: progress
  }));
};

export default Dialog;
