// @uploadcare/react-widget/en/uploader-428fc4eb.js
import { _ as _extends, a as _objectWithoutProperties, b as _slicedToArray, c as _objectSpread2, d as _defineProperty } from './index-c521bff0.js';
import React, { useRef, useEffect, useCallback, useImperativeHandle } from 'react';
import '@uploadcare/client-suspense';
import { u as useCustomTabs } from './use-custom-tabs-7f831c88.js';
import isEqual from 'react-fast-compare';
import uploadcare from 'uploadcare-widget/uploadcare.lang.en';

var useCommitedRef = function useCommitedRef(value) {
  var ref = useRef(value);
  useEffect(function () {
    ref.current = value;
  }, [value]);
  return ref;
};

var useEventCallback = function useEventCallback(callback) {
  var ref = useCommitedRef(callback);
  return useCallback(function () {
    return ref.current && ref.current.apply(ref, arguments);
  }, [ref]);
};

var useValidators = function useValidators(widget, validators) {
  useEffect(function () {
    if (validators != null) {
      var instance = widget.current;
      validators.forEach(function (validator) {
        instance.validators.push(validator);
      });
      return function () {
        instance.validators.length = 0;
      };
    }
  }, [widget, validators]);
};

function useDeepMemo(factory, deps) {
  var isValid = true;
  var valueRef = useRef(); // initial hook call

  if (!valueRef.current) {
    valueRef.current = {
      deps: deps,
      result: factory()
    }; // subsequent calls
  } else {
    isValid = !!(deps && valueRef.current.deps && isEqual(deps, valueRef.current.deps));
  }

  var cache = isValid ? valueRef.current : {
    deps: deps,
    result: factory()
  }; // must update immediately so any sync renders here don't cause an infinite loop

  valueRef.current = cache;
  return cache.result;
}

function camelCaseToDash(str) {
  return str.replace(/([a-zA-Z])(?=[A-Z])/g, '$1-').toLowerCase();
}

var propsToAttr = function propsToAttr(props) {
  return Object.entries(props).reduce(function (attr, _ref) {
    var _ref2 = _slicedToArray(_ref, 2),
        key = _ref2[0],
        value = _ref2[1];

    return _objectSpread2(_objectSpread2({}, attr), {}, _defineProperty({}, "data-".concat(camelCaseToDash(key)), value));
  }, {});
};

var useWidget = function useWidget(_ref3, uploadcare) {
  var id = _ref3.id,
      name = _ref3.name,
      value = _ref3.value,
      onFileSelect = _ref3.onFileSelect,
      onChange = _ref3.onChange,
      onDialogOpen = _ref3.onDialogOpen,
      onDialogClose = _ref3.onDialogClose,
      onTabChange = _ref3.onTabChange,
      apiRef = _ref3.apiRef,
      customTabs = _ref3.customTabs,
      validators = _ref3.validators,
      tabsCss = _ref3.tabsCss,
      options = _objectWithoutProperties(_ref3, ["id", "name", "value", "onFileSelect", "onChange", "onDialogOpen", "onDialogClose", "onTabChange", "apiRef", "customTabs", "validators", "tabsCss"]);

  var input = useRef(null);
  var widget = useRef(null);
  var fileSelectedCallback = useEventCallback(onFileSelect);
  var changeCallback = useEventCallback(onChange);
  var dialogOpenCallback = useEventCallback(onDialogOpen);
  var dialogCloseCallback = useEventCallback(onDialogClose);
  var tabChangeCallback = useEventCallback(onTabChange);
  useCustomTabs(customTabs, uploadcare);
  var attributes = useDeepMemo(function () {
    return propsToAttr(options);
  }, [options]);
  useEffect(function () {
    widget.current = uploadcare.Widget(input.current);
    var widgetElement = input.current.nextSibling;
    return function () {
      return widgetElement && widgetElement.remove();
    };
  }, [uploadcare, attributes]);
  useValidators(widget, validators);
  useEffect(function () {
    widget.current.onUploadComplete.add(changeCallback);
    widget.current.onChange.add(fileSelectedCallback);
    return function () {
      widget.current.onUploadComplete.remove(changeCallback);
      widget.current.onChange.remove(fileSelectedCallback);
    };
  }, [changeCallback, fileSelectedCallback]);
  useEffect(function () {
    var dialog;

    var saveDialog = function saveDialog(ref) {
      dialog = ref;
      dialog.done(dialogCloseCallback).fail(dialogCloseCallback).progress(tabChangeCallback);
      dialogOpenCallback(ref);
    };

    widget.current.onDialogOpen.add(saveDialog);
    return function () {
      widget.current.onDialogOpen.remove(saveDialog);
      dialog && dialog.reject();
    };
  }, [attributes, dialogCloseCallback, dialogOpenCallback, tabChangeCallback]);
  useEffect(function () {
    var files = [];

    var saveFiles = function saveFiles(file) {
      if (file) {
        files = file.files ? file.files() : [file];
      } else {
        files = [];
      }
    };

    widget.current.onChange.add(saveFiles);
    return function () {
      files.forEach(function (file) {
        return file.cancel();
      });
      widget.current.onChange.remove(saveFiles);
    };
  }, [attributes]);
  useEffect(function () {
    widget.current.value(value);
  }, [value]);
  useEffect(function () {
    if (uploadcare && tabsCss && typeof tabsCss === 'string') {
      if (tabsCss.indexOf('https://') === 0) {
        uploadcare.tabsCss.addUrl(tabsCss);
      } else {
        uploadcare.tabsCss.addStyle(tabsCss);
      }
    }
  }, [uploadcare, tabsCss]);
  useImperativeHandle(apiRef, function () {
    return {
      openDialog: function openDialog() {
        return widget.current.openDialog();
      },
      reloadInfo: function reloadInfo() {
        return widget.current.reloadInfo();
      },
      getInput: function getInput() {
        return widget.current.inputElement;
      }
    };
  }, []);
  return useCallback(function () {
    return /*#__PURE__*/React.createElement("input", _extends({
      type: "hidden",
      ref: input,
      id: id,
      name: name
    }, attributes));
  }, [attributes, id, name]);
};

var Uploader = function Uploader(props) {
  var Input = useWidget(props, uploadcare);
  return /*#__PURE__*/React.createElement(Input, null);
};

export default Uploader;
